2024-03-16T16:15:14.5899420Z /Users/runner/work/_temp/mariadb-11.4.1/storage/innobase/fil/fil0fil.cc:1415:15: error: use of undeclared identifier 'OS_DATA_FILE_NO_O_DIRECT'
2024-03-16T16:15:14.6001560Z         type= OS_DATA_FILE_NO_O_DIRECT;
2024-03-16T16:15:14.6101960Z               ^
table/block_based/data_block_footer.cc.o
2024-03-16T16:15:15.0131160Z 1 error generated.

diff --git a/storage/innobase/fil/fil0fil.cc b/storage/innobase/fil/fil0fil.cc
index 0ce54df6574..234daea5599 100644
--- a/storage/innobase/fil/fil0fil.cc
+++ b/storage/innobase/fil/fil0fil.cc
@@ -1410,10 +1410,12 @@ ATTRIBUTE_COLD void fil_space_t::reopen_all()
 
       ulint type= OS_DATA_FILE;
 
+#if defined _WIN32 || defined HAVE_FCNTL_DIRECT
       switch (FSP_FLAGS_GET_ZIP_SSIZE(space.flags)) {
       case 1: case 2:
         type= OS_DATA_FILE_NO_O_DIRECT;
       }
+#endif
 
       for (ulint count= 10000; count--;)
       {


2024-03-16T15:01:00.5179990Z /Users/runner/work/_temp/mariadb-10.11.7/extra/mariabackup/xtrabackup.cc:2431:39: error: use of undeclared identifier 'OS_DATA_FILE_NO_O_DIRECT'
2024-03-16T15:01:00.5282430Z                                       OS_DATA_FILE_NO_O_DIRECT, false, &ret);
2024-03-16T15:01:00.5385590Z                                       ^
2024-03-16T15:01:01.4599970Z 1 error generated.

diff --git a/extra/mariabackup/xtrabackup.cc b/extra/mariabackup/xtrabackup.cc
index 1482c5f3507..8379a51ea5a 100644
--- a/extra/mariabackup/xtrabackup.cc
+++ b/extra/mariabackup/xtrabackup.cc
@@ -2418,7 +2418,12 @@ static bool innodb_init()
   os_file_delete_if_exists_func(ib_logfile0.c_str(), nullptr);
   os_file_t file= os_file_create_func(ib_logfile0.c_str(),
                                       OS_FILE_CREATE, OS_FILE_NORMAL,
-                                      OS_DATA_FILE_NO_O_DIRECT, false, &ret);
+#if defined _WIN32 || defined HAVE_FCNTL_DIRECT
+                                      OS_DATA_FILE_NO_O_DIRECT,
+#else
+                                      OS_DATA_FILE,
+#endif
+                                      false, &ret);
   if (!ret)
   {
   invalid_log:
